### PR TITLE
Move Texture2DAtlas to core.

### PR DESCRIPTION
**Pull request by eh-unity**
This is a prototype draft!

* Add a prototype core asset class.
* Move blitting functions into a blitter class
* Move TextureHash to core.
* Change HDUtils.Blit into wrappers for core.Blitter.

# **Please read the [Contributing guide](CONTRIBUTING.md) before making a PR.**

* Read the [Graphics repository & Yamato FAQ](http://go/graphics-yamato-faq).

### Checklist for PR maker
- [ ] Have you added a backport label (if needed)? For example, the `need-backport-*` label. After you backport the PR, the label changes to `backported-*`.
- [ ] Have you updated the changelog? Each package has a `CHANGELOG.md` file.
- [ ] Have you updated or added the documentation for your PR? When you add a new feature, change a property name, or change the behavior of a feature, it's best practice to include related documentation changes in the same PR. If you do add documentation, make sure to add the relevant Graphics Docs team member as a reviewer of the PR. If you are not sure which person to add, see the [Docs team contacts sheet](https://docs.google.com/spreadsheets/d/1rgUWWgwLFEHIQ3Rz-LnK6PAKmbM49DZZ9al4hvnztOo/edit#gid=1058860420).
- [ ] Have you added a graphic test for your PR (if needed)? When you add a new feature, or discover a bug that tests don't cover, please add a graphic test.

---
### Purpose of this PR
Share generic graphics code classes and resources between different render-pipelines.
Share TextureAtlas code.

---
### Testing status
Describe what manual/automated tests were performed for this PR

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

**Review by JulienIgnace-Unity — APPROVED** [review]
Looks good to me.

File: com.unity.render-pipelines.core/Runtime/Textures/PowerOfTwoTextureAtlas.cs [hunk 1]
```diff
@@ -1,12 +1,9 @@
...
 using System.Collections.Generic;
-using UnityEngine;
-using UnityEngine.Rendering;
 using UnityEngine.Experimental.Rendering;
 
-namespace UnityEngine.Rendering.HighDefinition
+namespace UnityEngine.Rendering
 {
-    class PowerOfTwoTextureAtlas : Texture2DAtlas
+    public class PowerOfTwoTextureAtlas : Texture2DAtlas
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.core/Runtime/Textures/PowerOfTwoTextureAtlas.cs:6]: note: you will need to add documentation for it if we switch to public

File: com.unity.render-pipelines.core/Runtime/Utilities/Blitter.cs [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+        /// <param name="material">Material to invoke when blitting.</param>
+        /// <param name="pass">Pass idx within the material to invoke.</param>
+        public static void BlitTexture(CommandBuffer cmd, RTHandle source, Vector4 scaleBias, Material material, int pass)
+        {
+            s_PropertyBlock.SetVector(BlitShaderIDs._BlitScaleBias, scaleBias);
+            s_PropertyBlock.SetTexture(BlitShaderIDs._BlitTexture, source);
+            cmd.DrawProcedural(Matrix4x4.identity, material, pass, MeshTopology.Triangles, 3, 1, s_PropertyBlock);
+        }
+
+        // In the context of HDRP, the internal render targets used during the render loop are the same for all cameras, no matter the size of the camera.
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.core/Runtime/Utilities/Blitter.cs:128]: we will need to update the comment :)

File: com.unity.render-pipelines.core/Runtime/Utilities/Blitter.cs [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+        /// <param name="cmd">Command Buffer used for rendering.</param>
+        /// <param name="source">Source RTHandle.</param>
+        /// <param name="scaleBias">Scale and bias for sampling the input texture.</param>
+        /// <param name="material">Material to invoke when blitting.</param>
+        /// <param name="pass">Pass idx within the material to invoke.</param>
+        public static void BlitTexture(CommandBuffer cmd, RTHandle source, Vector4 scaleBias, Material material, int pass)
+        {
+            s_PropertyBlock.SetVector(BlitShaderIDs._BlitScaleBias, scaleBias);
+            s_PropertyBlock.SetTexture(BlitShaderIDs._BlitTexture, source);
+            cmd.DrawProcedural(Matrix4x4.identity, material, pass, MeshTopology.Triangles, 3, 1, s_PropertyBlock);
```
Comment (phi-lira) [inline com.unity.render-pipelines.core/Runtime/Utilities/Blitter.cs:125]: This depends on SV_VertexID, seems like that is not supported on GLES2 ?
https://docs.unity3d.com/Manual/SL-ShaderSemantics.html

**Review by phi-lira — CHANGES_REQUESTED** [review]
LGTM. The only thing I'm concerned is regarding support for GLES2.

File: com.unity.render-pipelines.core/Runtime/Utilities/Blitter.cs [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+using UnityEngine.Experimental.Rendering;
+using UnityEngine.Assertions;
+using System.Text.RegularExpressions;
+#if UNITY_EDITOR
+using UnityEditor;
+#endif
+
+namespace UnityEngine.Rendering
+{
+    public static class Blitter
```
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.core/Runtime/Utilities/Blitter.cs:15]: done.

File: com.unity.render-pipelines.core/Runtime/Utilities/Blitter.cs [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+        /// <summary>
+        /// Blit a RTHandle to another RTHandle.
+        /// This will properly account for partial usage (in term of resolution) of the texture for the current viewport.
+        /// </summary>
+        /// <param name="cmd">Command Buffer used for rendering.</param>
+        /// <param name="source">Source RTHandle.</param>
+        /// <param name="destination">Destination RTHandle.</param>
+        /// <param name="mipLevel">Mip level to blit.</param>
+        /// <param name="bilinear">Enable bilinear filtering.</param>
+        public static void BlitCameraTexture(CommandBuffer cmd, RTHandle source, RTHandle destination, float mipLevel = 0.0f, bool bilinear = false)
```
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.core/Runtime/Utilities/Blitter.cs:146]: What would be a better name?

File: com.unity.render-pipelines.core/Runtime/Utilities/Blitter.cs (hunk 3 above)
Comment (eh-unity) [reply com.unity.render-pipelines.core/Runtime/Utilities/Blitter.cs:134]: I think we could fix this with another integration PR.

File: com.unity.render-pipelines.core/Runtime/Textures/PowerOfTwoTextureAtlas.cs (hunk 1 above)
Comment (eh-unity) [reply com.unity.render-pipelines.core/Runtime/Textures/PowerOfTwoTextureAtlas.cs:9]: done.

File: com.unity.render-pipelines.core/Runtime/Textures/Texture2DAtlas.cs [hunk 6]
```diff
@@ -131,7 +130,7 @@ public void Reset()
         }
     }
 
-    class Texture2DAtlas
+    public class Texture2DAtlas
```
[the remark this answers is not in the document]
Comment (eh-unity) [reply com.unity.render-pipelines.core/Runtime/Textures/Texture2DAtlas.cs:138]: done.

**Review by phi-lira — APPROVED** [review]

**Comment by phi-lira** [comment]
Failures not related to this PR.

**Pull request merged**